File: 00_Core_Interactive_Showcase/frontend_web/src/features/rag-mission/services/ragService.js
import { qingyuCafeDocument } from "../data/ragDocument.js";

const delay = (ms = 720) => new Promise((resolve) => setTimeout(resolve, ms));

const tokenize = (text) =>
  text
    .toLowerCase()
    .replace(/[？?，,。.!！、]/g, " ")
    .split(/\s+/)
    .map((token) => token.trim())
    .filter(Boolean);

function scoreChunk(question, chunk) {
  const normalizedQuestion = question.toLowerCase();
  const questionTokens = tokenize(question);
  const keywordScore = chunk.keywords.reduce((score, keyword) => {
    return normalizedQuestion.includes(keyword.toLowerCase()) ? score + 3 : score;
  }, 0);
  const contentScore = questionTokens.reduce((score, token) => {
    return `${chunk.heading} ${chunk.content}`.toLowerCase().includes(token) ? score + 1 : score;
  }, 0);
  return keywordScore + contentScore;
}

function retrieveRelevantChunks(question) {
  const scored = qingyuCafeDocument.chunks
    .map((chunk) => ({
      ...chunk,
      score: scoreChunk(question, chunk),
    }))
    .sort((a, b) => b.score - a.score);

  const matched = scored.filter((chunk) => chunk.score > 0);
  const fallback = scored.filter((chunk) => chunk.score === 0);
  return [...matched, ...fallback].slice(0, 2);
}

function buildAnswer(question, chunks) {
  const topics = chunks.map((chunk) => chunk.heading).join("、");
  const hasStrongMatch = chunks[0]?.score >= 3;

  if (!hasStrongMatch) {
    return {
      answer:
        `我在「${qingyuCafeDocument.title}」中沒有找到完全對應「${question}」的明確片段。` +
        `目前最接近的來源是 ${topics}，建議補充文件或新增 FAQ 後再回答。`,
      confidence: "low",
    };
  }

  return {
    answer:
      `根據「${qingyuCafeDocument.title}」的 ${topics} 片段，` +
      chunks.map((chunk) => chunk.content).join(" "),
    confidence: chunks.length >= 2 ? "high" : "medium",
  };
}

async function ask(question) {
  await delay();
  const retrievedChunks = retrieveRelevantChunks(question);
  const answerPayload = buildAnswer(question, retrievedChunks);

  return {
    question,
    answer: answerPayload.answer,
    confidence: answerPayload.confidence,
    document: {
      id: qingyuCafeDocument.id,
      title: qingyuCafeDocument.title,
    },
    sources: retrievedChunks.map((chunk) => ({
      id: chunk.id,
      sourceId: chunk.sourceId,
      heading: chunk.heading,
      snippet: chunk.content,
      score: chunk.score,
      fullText: `${chunk.heading}\n${chunk.content}`,
    })),
    citedSnippets: retrievedChunks.map((chunk) => ({
      sourceId: chunk.sourceId,
      text: chunk.content,
    })),
    retrievalTrace: {
      method: "mock keyword scoring",
      topK: retrievedChunks.length,
      note: "Replace this mock service with a FastAPI RAG endpoint later.",
    },
  };
}

export const ragService = {
  ask,
};

File: 00_Core_Interactive_Showcase/backend_api/app/services/rag_service.py
from app.schemas.rag import RagAskRequest, RagAskResponse, RagSource


MOCK_SOURCES = [
    RagSource(
        source_id="Source 1",
        heading="營業時間",
        snippet="晴宇咖啡週一到週五 10:00-20:00 營業，週末與國定假日 09:00-21:00。",
    ),
    RagSource(
        source_id="Source 2",
        heading="預約方式",
        snippet="顧客可以透過 LINE 官方帳號預約座位。六人以上建議提前一天預約。",
    ),
]


def ask_rag(payload: RagAskRequest) -> RagAskResponse:
    answer = (
        f"根據文件 `{payload.document_id}` 的 mock retrieval 結果，"
        f"針對「{payload.question}」最相關的來源是營業時間與預約方式。"
    )
    return RagAskResponse(
        answer=answer,
        confidence="medium",
        sources=MOCK_SOURCES,
        cited_snippets=[source.snippet for source in MOCK_SOURCES],
    )
